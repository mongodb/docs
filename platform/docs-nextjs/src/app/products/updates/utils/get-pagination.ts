/**
 *
 * @param array - flatten array
 * @param chunkSize - max size of each array
 * @returns [[{}, {}], [{}]]
 * @description Takes an flatten array and creates one level nested array of objects based on the chunkSize. Used for pagination data.
 */
export const getPagination = <T>(array: T[], chunkSize: number): T[][] => {
  const chunks: T[][] = [];
  if (!array.length) {
    return [[]]; // default to one empty nested array
  }
  for (let i = 0; i < array.length; i += chunkSize) {
    chunks.push(array.slice(i, i + chunkSize));
  }
  return chunks;
};
