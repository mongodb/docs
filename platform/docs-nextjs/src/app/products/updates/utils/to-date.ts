export const normalizeDate = (date: string) => {
  // Check if the date string has the correct format 'YYYY-MM-DD'
  const dateToNormalize = /^(\d{4})-(\d{2})-(\d{2})$/.exec(date)?.slice(1); // excludes the full date in the first index i.e. "2025-05-27"
  if (dateToNormalize) {
    // This is needed because, JavaScript’s Date constructor uses zero-based months (January = 0 … December = 11).
    return new Date(Number(dateToNormalize[0]), Number(dateToNormalize[1]) - 1, Number(dateToNormalize[2]));
  }
  return new Date(date);
};
