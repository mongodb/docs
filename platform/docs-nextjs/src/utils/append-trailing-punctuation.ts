import { ASTNode, ReferenceNode, RefRoleNode } from "@/types/ast";
import { isReferenceNode, isRefRoleNode, isTextNode } from "@/types/ast-utils";

// Append dangling punctuation to preceding Link's text
export const appendTrailingPunctuation = (nodes: ASTNode[]) => {
  const truncatedNodes = [];

  for (let i = 0; i < nodes.length; i++) {
    const currNode = nodes[i];
    const nextNode = nodes[i + 1];
    const hasDanglingSibling =
      isTextNode(nextNode) &&
      nextNode.value.length === 1 &&
      !nextNode.value.match(/\s/);
    // TODO: if issues arise, instead of matching for non-whitespace, we might consider checking FOR certain singular punctuation
    // ie: regex match for  .?!,:;)*"']}`

    if ((isReferenceNode(currNode) || isRefRoleNode(currNode)) && hasDanglingSibling) {
      const copyOfNodeWithDeepCopiedChildren = deepCopyNodesChildren(currNode);
      copyOfNodeWithDeepCopiedChildren.children.push(nextNode);
      truncatedNodes.push(copyOfNodeWithDeepCopiedChildren);
      i++;
      continue;
    }
    truncatedNodes.push(currNode);
  }
  return truncatedNodes;
};

// Make a copy of node with a deep copy of the new child node
function deepCopyNodesChildren(refNode: ReferenceNode | RefRoleNode) {
  if (!refNode.children || !refNode.children.length) return { ...refNode };
  const copyOfChildren = [...refNode.children];

  const copyOfNode = {
    ...refNode,
    children: copyOfChildren,
  };
  return copyOfNode;
}
