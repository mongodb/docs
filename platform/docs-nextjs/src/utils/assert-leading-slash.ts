const assertLeadingSlash = (url: string) => {
  if (url && url.match(/^\//)) {
    return url;
  }
  return `/${url}`;
};

export { assertLeadingSlash };
