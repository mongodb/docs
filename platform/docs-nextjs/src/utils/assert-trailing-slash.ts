export const assertTrailingSlash = (url: string) => {
  if (url && url.match(/\/$/)) {
    return url;
  }
  return `${url}/`;
};
