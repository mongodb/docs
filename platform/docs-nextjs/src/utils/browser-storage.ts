import { isBrowser } from './is-browser';

const isValidStorage = isBrowser;

export const setLocalValue = (key: string, value: unknown) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      localStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting localStorage value');
  }
};

export const getLocalValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.localStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting localStorage value');
    return undefined;
  }
};

export const setSessionValue = (key: string, value: string) => {
  try {
    if (isValidStorage) {
      const prevState = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      sessionStorage.setItem('mongodb-docs', JSON.stringify({ ...prevState, [key]: value }));
    }
  } catch {
    console.error('Error setting session value');
  }
};

export const getSessionValue = (key: string) => {
  try {
    if (isValidStorage && JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}')) {
      const docsObj = JSON.parse(window.sessionStorage.getItem('mongodb-docs') || '{}');
      if (docsObj) {
        return docsObj[key];
      }
    }
    return undefined;
  } catch {
    console.error('Error getting session value');
    return undefined;
  }
};
