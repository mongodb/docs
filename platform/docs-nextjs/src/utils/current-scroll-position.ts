import { isBrowser } from './is-browser';

// Function that returns current scroll position for analytics
export const currentScrollPosition = (): number => {
  if (isBrowser) {
    const percentage = Math.round((window.scrollY / (document.body.scrollHeight - window.innerHeight)) * 100);
    return Math.round(percentage / 10) * 10;
  }
  return 0;
};
