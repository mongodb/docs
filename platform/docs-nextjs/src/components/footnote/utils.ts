import type { ASTNode } from '@/types/ast';
import type { FootnoteNode, FootnoteReferenceNode } from '@/types/ast';
import { findAllKeyValuePairs } from '@/utils/find-all-key-value-pairs';
import type { FootnoteType } from './footnote-context';

export const getFootnotes = (nodes: ASTNode[]) => {
  const footnotes = findAllKeyValuePairs(nodes, 'type', 'footnote') as FootnoteNode[];
  const footnoteReferences = findAllKeyValuePairs(nodes, 'type', 'footnote_reference') as FootnoteReferenceNode[];
  const numAnonRefs = footnoteReferences.filter(
    (node) => !Object.prototype.hasOwnProperty.call(node, 'refname'),
  ).length;
  // We label our footnotes by their index, regardless of their names to
  // circumvent cases such as [[1], [#], [2], ...]
  return footnotes.reduce<Record<string, FootnoteType>>((map, footnote, index) => {
    if (footnote.name) {
      // Find references associated with a named footnote

      map[footnote.name] = {
        label: index + 1,
        references: getNamedFootnoteReferences(footnoteReferences, footnote.name),
      };
    } else {
      // Find references associated with an anonymous footnote
      // Replace potentially broken anonymous footnote ids
      footnote.id = `${index + 1}`;

      map[footnote.id] = {
        label: index + 1,
        references: getAnonymousFootnoteReferences(index, numAnonRefs),
      };
    }
    return map;
  }, {});
};

// Find all footnote_reference node IDs associated with a given footnote by
// that footnote's refname
const getNamedFootnoteReferences = (footnoteReferences: FootnoteReferenceNode[], refname: string) => {
  return footnoteReferences.filter((node) => node.refname === refname).map((node) => node.id);
};

// They are used infrequently, but here we match an anonymous footnote to its reference.
// The nth footnote on a page is associated with the nth reference on the page. Since
// anon footnotes and footnote references are anonymous, we assume a 1:1 pairing, and
// have no need to query nodes. If there are more anonymous footnotes than references,
// we may return an empty array
const getAnonymousFootnoteReferences = (index: number, numAnonRefs: number) => {
  return index > numAnonRefs ? [] : [`id${index + 1}`];
};
