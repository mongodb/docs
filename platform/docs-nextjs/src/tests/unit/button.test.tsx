import { render } from '@testing-library/react';
import { mockLocation } from '../utils/mock-location';
import Button from '@/components/button';

import mockData from '@/tests/data/button.test.json';
import type { ButtonNode } from '@/types/ast';

beforeAll(() => {
  mockLocation({ hash: '' });
});

describe('button component', () => {
  it('renders correctly', () => {
    const typedMockData = mockData as ButtonNode;
    const tree = render(<Button options={typedMockData.options} argument={typedMockData.argument} darkMode={false} />);
    // expect anchor tag to be generated from uri option
    expect(tree.container.querySelector('a')).toBeInTheDocument();
    expect(tree.asFragment()).toMatchSnapshot();
  });
});
