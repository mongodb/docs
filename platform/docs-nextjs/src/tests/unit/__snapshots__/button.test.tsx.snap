// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`button component renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: 1px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  z-index: 0;
  font-family: 'Euclid Circular A','Helvetica Neue',Helvetica,Arial,sans-serif;
  border-radius: 6px;
  background-color: #00684A;
  border-color: #00684A;
  color: #FFFFFF;
  font-size: 13px;
  line-height: 20px;
  font-weight: 500;
  height: 36px;
}

.emotion-0:focus-visible,
.emotion-0[data-focus="true"] {
  outline: none;
}

.emotion-0:active,
.emotion-0[data-active="true"],
.emotion-0:focus,
.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:focus-visible,
.emotion-0[data-focus="true"] {
  color: #FFFFFF;
}

.emotion-0:hover,
.emotion-0[data-hover="true"],
.emotion-0:active,
.emotion-0[data-active="true"] {
  color: #FFFFFF;
  background-color: #00593F;
  border-color: #00593F;
  box-shadow: 0 0 0 3px #C0FAE6;
}

.emotion-0:focus-visible,
.emotion-0[data-focus="true"] {
  color: #FFFFFF;
  background-color: #00593F;
  box-shadow: 0 0 0 2px #FFFFFF,0 0 0 4px #0498EC;
}

.emotion-1 {
  overflow: hidden;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 5px;
}

.emotion-1 {
  overflow: hidden;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 5px;
}

.emotion-2 {
  display: grid;
  grid-auto-flow: column;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  z-index: 0;
  -webkit-transition: all 150 ease-in-out;
  transition: all 150 ease-in-out;
  padding: 0 11px;
  gap: 6px;
}

<a
    aria-disabled="false"
    class="lg-ui-button-0000 emotion-0 button"
    data-lgid="lg-button"
    href="/install"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    >
      Download Compass
    </div>
  </a>
</DocumentFragment>
`;
